README.en

```diff
@@ -0,0 +1,357 @@
+
+Encodings
+=========
+
+
+Pire supports Latin-1 and UTF-8 encodings by default. The required instances of the Pire::Encoding classes can be
+obtained by calling the appropriate functions from the namespace Pire::Encodings.
+
+Support for other encodings inheriting from Pire::Encoding can be added.
+In the inherited class, the following methods should be overridden:
+
+    * wchar32 FromLocal(const char*& begin, const char* end) reads and returns input character and forwards begin.
+      Returns an exceprion if the input includes an invalid sequence.
+    * std::string ToLocal(wchar32) returns returns the byte representation of the character in this encoding.
+      If the character in this encoding is not representable, returns an empty string.
+    * AppendDot(Fsm&) adds a fragment to the end machine that allows any one character in this encoding.
+
+After that, an instance of such a class can be passed to lexer.SetEncoding() along with the built-in ones.
+
+Serialization, MMAP() AND Inlining
+=================================
+
+Built scanners can be saved to std::ostream or uploaded from std::istream by calling Scanner:: Save()
+and Scanner::Load() (similarly for other scanners).
+
+If the scanner was saved to a file, and then the file was mapped to memory, then
+instead of calling Scanner::Load(), Scanner:: Mmap() that will not copy the
+table with transitions can be used. Mmap() returns a pointer to the first byte
+after the regular expression is serialized.
+
+Note that the scanner beginning must be in memory at the address aligned on the boundary of the machine word.
+If something other than serialized scanners is written to the file (the scanner views themselves always takes
+an integer number of machine words), manual alignment or alignment with the AlignedInput and AlignedOutput classes
+should be performed.
+
+The serialized scanner is not portable between architectures (even between x86 and x86_64).
+When trying to read or map the regular expression serialized in another architecture,
+an exception will be returned.
+
+If the code requires using the same predetermined regular expression,
+static const Scanner = Pire::Lexer(«...»).Parse().Compile() can be used, but in this case
+compilation of the expression will be performed every time the program starts.
+To avoid delays in the program starting, the expression can be compiled into a scanner,
+serialized, substituted into the code as a string literal, and called Scanner::Mmap() on it.
+The pire_inline program performs all these actions. The program accepts the file with the C++ code,
+searches all the PIRE_REGEXP("pattern", "flags") entries outside the comments and string in the file
+and replaces found entries with an expression that returns the machine with a ready table of transitions.
+
+The following characters can be in the string with flags:
+
+    * i - case insensitive
+    * u - compile the expression in UTF-8
```

Comment:
The terms "regular expression" and "pattern" can be used interchangeably, but I would not use just "expression" as that is a broader term.